# Combines position specific (i) conservation, (ii) subfamily scores, (iii)
# SNP data, and (iv) cancer data

source("code/000_libraries.R")
source("code/000_functions.R")

################################################################################

# (1) CHEMOKINE ----------------------------------------------------------------
# chemokine import
ck.cons <- read_csv("data/sequence/chemokine/processed/CK_CONSERVATION.csv")
ck.cc.cxc <- read_csv("data/sequence/chemokine/processed/CK_LOGISTIC_REGRESSION_ACCURACY_N3.csv")
colnames(ck.cc.cxc) <- c("ccn", "cc_cxc_lr", "cc_cxc_lr_sd")
ck.snp <- read_csv("data/variant/gnomad/processed/CK_GNOMAD_COUNTS_FREQ.csv")
ck.can <- read_csv("data/variant/tcga/processed/CK_TCGA_COUNTS.csv")

# chemokine merge
ck.cons <- left_join(ck.cons,  ck.cc.cxc)
rm(ck.cc.cxc)
ck.cons <- ck.cons %>% separate(protein, into = c("protein", "b"), sep = "_") %>%
  dplyr::select(-b)
ck.cons$protein <- toupper(ck.cons$protein)
ck.snp.can <- left_join(ck.snp, ck.can)
rm(ck.can, ck.snp)
ck.master <- left_join(ck.cons, ck.snp.can) 
# note that the conservation df has 472 x 43 rows versus
# cancer/snp df with 472 x 46
rm(ck.cons, ck.snp.can)

# now add CC/CXC classification accuracy SCORE generated by each position of
# each chemokine
ck.cc.cxc.score <- read_csv("data/sequence/chemokine/processed/CK_PARALOG_LOGISTIC_REGRESSION_ACCURACY_N3.csv")
ck.cc.cxc.score$protein <- toupper(ck.cc.cxc.score$protein)
ck.cc.cxc.score <- ck.cc.cxc.score %>% dplyr::select(-class, -run, -score) %>% unique()
colnames(ck.cc.cxc.score) <- c("protein", "ccn", "cc_cxc_lr_score", "cc_cxc_lr_score_sd")
ck.master <- left_join(ck.master, ck.cc.cxc.score)
rm(ck.cc.cxc.score)
ck.master <- ck.master %>% dplyr::select(protein, ccn, dom, all_para, all_cc_cxc_para, cc_para, cxc_para,
                                         ortho_cons, ngaps, pctgaps, cc_cxc_lr, cc_cxc_lr_sd, cc_cxc_lr_score, cc_cxc_lr_score_sd,
                                         snp_count, snp_freq_count, cancer_mut_count)

# add sequence
aln <- Align2DataFrame("data/sequence/chemokine/alignments/ALL_para.fasta", 
                        "data/lookup/SEQUENCES_CK_ALL_ORTHO_PARA_NTERM_UPDATED.txt")
pname <- aln %>% separate(seq, "|", remove = TRUE) %>% dplyr::select(1) 
colnames(pname) <- c("protein")
aln <- cbind(pname, aln)
rm(pname)
aln <- aln %>% dplyr::select(-seq)
aln <- aln %>% pivot_longer(cols = (2:ncol(aln)), names_to = c("ccn"))
aln$protein <- toupper(aln$protein)
ck.master <- left_join(ck.master, aln)
rm(aln)

# write csv
# write_csv(ck.master, "data/integrated/CK_CONS_CCCXC_SNP_CAN.csv") # LAST WRITTEN 20230930
rm(ck.master)

# (2) RECEPTOR -----------------------------------------------------------------
# receptor
ckr.cons <- read_csv("data/sequence/gpcr/processed/CKR_CONSERVATION.csv")
ckr.cc.cxc <- read_csv("data/sequence/gpcr/processed/CKR_LOGISTIC_REGRESSION_ACCURACY_N3.csv")
colnames(ckr.cc.cxc) <- c("gn", "cc_cxc_lr", "cc_cxc_lr_sd")
ckr.snp <- read_csv("data/variant/gnomad/processed/CKR_GNOMAD_COUNTS_FREQ.csv")
ckr.can <- read_csv("data/variant/tcga/processed/CKR_TCGA_COUNTS.csv")

# receptor merge
ckr.cons <- left_join(ckr.cons,  ckr.cc.cxc)
rm(ckr.cc.cxc)
ckr.cons$protein <- toupper(ckr.cons$protein)
ckr.snp.can <- left_join(ckr.snp, ckr.can)
rm(ckr.can, ckr.snp)
ckr.master <- left_join(ckr.cons, ckr.snp.can) 
# note that all receptors have 23 x 453 rows
rm(ckr.cons, ckr.snp.can)

# now add CC/CXC classification accuracy SCORE generated by each position of
# each receptor
ckr.cc.cxc.score <- read_csv("data/sequence/gpcr/processed/CKR_CLASSA_LOGISTIC_REGRESSION_ACCURACY_N3.csv")
ckr.cc.cxc.score$protein <- toupper(ckr.cc.cxc.score$protein)
ckr.cc.cxc.score <- ckr.cc.cxc.score %>% dplyr::select(-cc_cxc, -class, -run) %>% unique()
colnames(ckr.cc.cxc.score) <- c("gn", "protein", "cc_cxc_lr_score", "cc_cxc_lr_score_sd")
ckr.master <- left_join(ckr.master, ckr.cc.cxc.score)
rm(ckr.cc.cxc.score)
ckr.master <- ckr.master %>% dplyr::select(protein, gn, dom, all_para, all_non_ackr_para, all_cc_cxc_para, all_classa, cc_para, cxc_para, ack_para,
                                    ortho_cons, ngaps, pctgaps, cc_cxc_lr, cc_cxc_lr_sd, cc_cxc_lr_score, cc_cxc_lr_score_sd,
                                    snp_count, snp_freq_count, cancer_mut_count)

# add sequence
aln <- Align2DataFrame("data/sequence/gpcr/alignments/ALL_para.fasta", 
                       "data/lookup/SEQUENCES_CKR_ALL_ORTHO_PARA_NTERM_ECL2_UPDATED.txt")
pname <- aln %>% separate(seq, "|", remove = TRUE) %>% dplyr::select(1) 
colnames(pname) <- c("protein")
aln <- cbind(pname, aln)
rm(pname)
aln <- aln %>% dplyr::select(-seq)
aln <- aln %>% pivot_longer(cols = (2:ncol(aln)), names_to = c("gn"))
aln$protein <- toupper(aln$protein)
ckr.master <- left_join(ckr.master, aln)
rm(aln)

# write csv
# write_csv(ckr.master, "data/integrated/CKR_CONS_CCCXC_SNP_CAN.csv") # LAST WRITTEN 20230930
rm(ckr.master)
