# Imports conservation scores generated by Mstatx and compiles them into a 
# single document. All Mstatx conservation calculations use alignments generated 
# from master alignment SEQUENCES_CKR_ALL_ORTHO_PARA_NTERM_ECL2_UPDATED.fasta
# located in folder /data/sequence/gpcr/alignments; CGN associated with alignments
# are in folder /data/lookup as SEQUENCES_CKR_ALL_ORTHO_PARA_NTERM_ECL2_UPDATED.txt
# Per position conservation scores by Mstatx were generated with the following 
# alignments sets (29 total):
# - Ortholog conservation (x23)
# - CC paralog conservation (1)
# - CXC paralog conservation (1)
# - CC and CXC paralog conservation (1)
# - ACK paralog conservation (1)
# - All non-ACKR para conservation (1)
# - All 23 paralog conservation (1)
# - All Class A GPCRs (1)
#
# Mstatx commands run using variations on the following command locally via
# command line terminal:
# mstatx -i ALL_para.fasta -s trident -m /Applications/MstatX-master/data/aaindex/HENS920102.mat -o ALL_para.txt
#
# Validated that scores are accurately assigned to columns via N-term spot
# checks - checked 0's appropriately assigned in data frame based on gapped
# columns from the alignment prior to the start of a given sequence
#
# Note on Class A GPCR alignment:
# (1) Class A GPCRs downloaded from GPCRdb.org
# (2) N-terminal Cys: For the following receptors which **have Cys at 7x24** 
#   and thus likely to have conserved 7x24-1x22 disulfide, adjusted N-terminal 
#   Cys to align to 1x22 position (51st residue position in alignment)
#   agrt1, ap1, bkbr1, bkbr2, ednra, ednrb, cltr1, lpar4, lpar6, gpr55, p2y1,
#   p2y2, p2y4, p2y6, hcar2, hcar3, oxgr1, gpr4, gpr17, gpr25, gpr34, gpr132,
#   gpr171, gpr174, gpr182, gpr183, p2y10
# - For all of above moved 1x23-1x30 positions to approximate 1x22 (ie removed
#   intervening gaps); 1x30 is unaffected in all seqs
# - For the following which have Cys near 7x24 (likely poor alignment), also
#   adjusted Nterm Cys to 1x22 position: cltr2, hcar1, sucr1
# (3) ECL2: Selected gnECL2.1:gnECL2.Cp11 > Removed gaps after Cys so all seqs 
#   left adjusted to Cys; Exceptions include GPCRs for which 4x50 is not Cys 
#   (eg lpar, s1pr) which have different ECL2 architecture
#
# Note on ECL2 conservation scoring:
# Note that no adjustments were made to GPCRdb numbering in ECL2 PRIOR to 45x50
# (see above for adjustments to GPCRdb numbering made AFTER 45x50). In GPCRdb
# numbering, residues prior to 45x50 are right-adjusted to TM4. As such,
# positions in region after TM4 but prior to 45x50 are NOT structurally equivalent
# and hence no conservation scoring is reported for these positions

source("code/000_libraries.R")
source("code/000_functions.R")

################################################################################

# (1) IMPORT AND TIDY ----------------------------------------------------------
# import scores
all.para <- read.table("data/sequence/gpcr/mstatx/ALL_para.txt", header = FALSE)
all.para <- all.para %>% select(-V1)
colnames(all.para) <- c("all_para")

all.non.ackr.para <- read.table("data/sequence/gpcr/mstatx/ALL_non_ackr_para.txt", header = FALSE)
all.non.ackr.para <- all.non.ackr.para %>% select(-V1)
colnames(all.non.ackr.para) <- c("all_non_ackr_para")

all.cc.cxc.para <- read.table("data/sequence/gpcr/mstatx/ALL_CC_CXC_para.txt", header = FALSE)
all.cc.cxc.para <- all.cc.cxc.para %>% select(-V1)
colnames(all.cc.cxc.para) <- c("all_cc_cxc_para")

cc.para <- read.table("data/sequence/gpcr/mstatx/CC_para.txt", header = FALSE)
cc.para <- cc.para %>% select(-V1)
colnames(cc.para) <- c("cc_para")

cxc.para <- read.table("data/sequence/gpcr/mstatx/CXC_para.txt", header = FALSE)
cxc.para <- cxc.para %>% select(-V1)
colnames(cxc.para) <- c("cxc_para")

ack.para <- read.table("data/sequence/gpcr/mstatx/ACK_para.txt", header = FALSE)
ack.para <- ack.para %>% select(-V1)
colnames(ack.para) <- c("ack_para")

all.classa <- read.table("data/sequence/gpcr/mstatx/ALL_classa.txt", header = FALSE)
all.classa <- all.classa %>% select(-V1)
colnames(all.classa) <- c("all_classa")

ccr1.ortho <- read.table("data/sequence/gpcr/mstatx/ccr1_ortho.txt", header = FALSE)
ccr1.ortho <- ccr1.ortho %>% select(-V1)
colnames(ccr1.ortho) <- c("ccr1")

ccr2.ortho <- read.table("data/sequence/gpcr/mstatx/ccr2_ortho.txt", header = FALSE)
ccr2.ortho <- ccr2.ortho %>% select(-V1)
colnames(ccr2.ortho) <- c("ccr2")

ccr3.ortho <- read.table("data/sequence/gpcr/mstatx/ccr3_ortho.txt", header = FALSE)
ccr3.ortho <- ccr3.ortho %>% select(-V1)
colnames(ccr3.ortho) <- c("ccr3")

ccr4.ortho <- read.table("data/sequence/gpcr/mstatx/ccr4_ortho.txt", header = FALSE)
ccr4.ortho <- ccr4.ortho %>% select(-V1)
colnames(ccr4.ortho) <- c("ccr4")

ccr5.ortho <- read.table("data/sequence/gpcr/mstatx/ccr5_ortho.txt", header = FALSE)
ccr5.ortho <- ccr5.ortho %>% select(-V1)
colnames(ccr5.ortho) <- c("ccr5")

ccr6.ortho <- read.table("data/sequence/gpcr/mstatx/ccr6_ortho.txt", header = FALSE)
ccr6.ortho <- ccr6.ortho %>% select(-V1)
colnames(ccr6.ortho) <- c("ccr6")

ccr7.ortho <- read.table("data/sequence/gpcr/mstatx/ccr7_ortho.txt", header = FALSE)
ccr7.ortho <- ccr7.ortho %>% select(-V1)
colnames(ccr7.ortho) <- c("ccr7")

ccr8.ortho <- read.table("data/sequence/gpcr/mstatx/ccr8_ortho.txt", header = FALSE)
ccr8.ortho <- ccr8.ortho %>% select(-V1)
colnames(ccr8.ortho) <- c("ccr8")

ccr9.ortho <- read.table("data/sequence/gpcr/mstatx/ccr9_ortho.txt", header = FALSE)
ccr9.ortho <- ccr9.ortho %>% select(-V1)
colnames(ccr9.ortho) <- c("ccr9")

ccr10.ortho <- read.table("data/sequence/gpcr/mstatx/ccr10_ortho.txt", header = FALSE)
ccr10.ortho <- ccr10.ortho %>% select(-V1)
colnames(ccr10.ortho) <- c("ccr10")

cxcr1.ortho <- read.table("data/sequence/gpcr/mstatx/cxcr1_ortho.txt", header = FALSE)
cxcr1.ortho <- cxcr1.ortho %>% select(-V1)
colnames(cxcr1.ortho) <- c("cxcr1")

cxcr2.ortho <- read.table("data/sequence/gpcr/mstatx/cxcr2_ortho.txt", header = FALSE)
cxcr2.ortho <- cxcr2.ortho %>% select(-V1)
colnames(cxcr2.ortho) <- c("cxcr2")

cxcr3.ortho <- read.table("data/sequence/gpcr/mstatx/cxcr3_ortho.txt", header = FALSE)
cxcr3.ortho <- cxcr3.ortho %>% select(-V1)
colnames(cxcr3.ortho) <- c("cxcr3")

cxcr4.ortho <- read.table("data/sequence/gpcr/mstatx/cxcr4_ortho.txt", header = FALSE)
cxcr4.ortho <- cxcr4.ortho %>% select(-V1)
colnames(cxcr4.ortho) <- c("cxcr4")

cxcr5.ortho <- read.table("data/sequence/gpcr/mstatx/cxcr5_ortho.txt", header = FALSE)
cxcr5.ortho <- cxcr5.ortho %>% select(-V1)
colnames(cxcr5.ortho) <- c("cxcr5")

cxcr6.ortho <- read.table("data/sequence/gpcr/mstatx/cxcr6_ortho.txt", header = FALSE)
cxcr6.ortho <- cxcr6.ortho %>% select(-V1)
colnames(cxcr6.ortho) <- c("cxcr6")

cx3cr1.ortho <- read.table("data/sequence/gpcr/mstatx/cx3cr1_ortho.txt", header = FALSE)
cx3cr1.ortho <- cx3cr1.ortho %>% select(-V1)
colnames(cx3cr1.ortho) <- c("cx3cr1")

xcr1.ortho <- read.table("data/sequence/gpcr/mstatx/xcr1_ortho.txt", header = FALSE)
xcr1.ortho <- xcr1.ortho %>% select(-V1)
colnames(xcr1.ortho) <- c("xcr1")

ackr1.ortho <- read.table("data/sequence/gpcr/mstatx/ackr1_ortho.txt", header = FALSE)
ackr1.ortho <- ackr1.ortho %>% select(-V1)
colnames(ackr1.ortho) <- c("ackr1")

ackr2.ortho <- read.table("data/sequence/gpcr/mstatx/ackr2_ortho.txt", header = FALSE)
ackr2.ortho <- ackr2.ortho %>% select(-V1)
colnames(ackr2.ortho) <- c("ackr2")

ackr3.ortho <- read.table("data/sequence/gpcr/mstatx/ackr3_ortho.txt", header = FALSE)
ackr3.ortho <- ackr3.ortho %>% select(-V1)
colnames(ackr3.ortho) <- c("ackr3")

ackr4.ortho <- read.table("data/sequence/gpcr/mstatx/ackr4_ortho.txt", header = FALSE)
ackr4.ortho <- ackr4.ortho %>% select(-V1)
colnames(ackr4.ortho) <- c("ackr4")

ccrl2.ortho <- read.table("data/sequence/gpcr/mstatx/ccrl2_ortho.txt", header = FALSE)
ccrl2.ortho <- ccrl2.ortho %>% select(-V1)
colnames(ccrl2.ortho) <- c("ccrl2")

# combine
data <- bind_cols(all.para, all.non.ackr.para, all.cc.cxc.para, all.classa, cc.para, cxc.para, ack.para,
                  ccr1.ortho, ccr2.ortho, ccr3.ortho, ccr4.ortho, ccr5.ortho, 
                  ccr6.ortho, ccr7.ortho, ccr8.ortho, ccr9.ortho, ccr10.ortho,
                  cxcr1.ortho, cxcr2.ortho, cxcr3.ortho, cxcr4.ortho,
                  cxcr5.ortho, cxcr6.ortho,
                  ackr1.ortho, ackr2.ortho, ackr3.ortho, ackr4.ortho, ccrl2.ortho,
                  cx3cr1.ortho, xcr1.ortho)

rm(all.para, all.non.ackr.para, all.cc.cxc.para, all.classa, cc.para, cxc.para, ack.para,
   ccr1.ortho, ccr2.ortho, ccr3.ortho, ccr4.ortho, ccr5.ortho, 
   ccr6.ortho, ccr7.ortho, ccr8.ortho, ccr9.ortho, ccr10.ortho,
   cxcr1.ortho, cxcr2.ortho, cxcr3.ortho, cxcr4.ortho,
   cxcr5.ortho, cxcr6.ortho,
   ackr1.ortho, ackr2.ortho, ackr3.ortho, ackr4.ortho, ccrl2.ortho,
   cx3cr1.ortho, xcr1.ortho)

# assign column names
aln.names <- c(read.table("data/lookup/SEQUENCES_CKR_ALL_ORTHO_PARA_NTERM_ECL2_UPDATED.txt", sep = ",", colClasses = "character"))
aln.names <- as.data.frame(aln.names)
aln.names <- t(aln.names)
aln.names <- as.data.frame(aln.names)
colnames(aln.names) <- c("gn")
data <- bind_cols(aln.names, data)
rownames(data) <- 1:nrow(data)
rm(aln.names)

# spread data
data <- data %>% gather(protein, ortho_cons, 9:ncol(data))


# (2) ADD ADDITIONAL DESCRIPTORS -----------------------------------------------
# add domain
lookup <- read_csv("data/lookup/lookup_gnccn_to_domain.csv")
lookup$gn <- c("gn")
lookup <- lookup %>% unique()
lookup <- lookup %>% unite(gn, c(gn, bwccn), sep = "")
data$dom <- lookup$dom[match(unlist(data$gn), lookup$gn)]
rm(lookup)

# add gap numbers (CKR orthologs; 951 seqs) and percentages
aln <- readAAMultipleAlignment("data/sequence/gpcr/alignments/SEQUENCES_CKR_ALL_ORTHO_PARA_NTERM_ECL2_UPDATED.fasta")
aln <- as.matrix(aln)
aln.names <- c(read.table("data/lookup/SEQUENCES_CKR_ALL_ORTHO_PARA_NTERM_ECL2_UPDATED.txt", sep = ",", colClasses = "character"))
colnames(aln) <- as.matrix(aln.names)
aln <- as.data.frame(aln)
#rm(aln.names)

gaps <- NULL
for (i in 1:ncol(aln)){
  temp <- as.data.frame(table(aln[,i]))
  colnames(temp) <- c("resno", "n")
  temp <- subset(temp, resno == "-")
  temp$pos <- i
  gaps <- rbind.data.frame(gaps, temp)
  rm(temp)
}
rm(i)
gaps$pct <- gaps$n/nrow(aln)
aln.names <- t(as.data.frame(aln.names))
colnames(aln.names) <- c("gn")
rm(aln)

gaps <- bind_cols(aln.names, gaps)
rm(aln.names)
colnames(gaps) <- c("gn", "restype", "n", "pos", "pct")
gaps <- select(gaps, gn, n, pct)
gaps$gn <- as.character(gaps$gn)

data <- left_join(data, gaps)
colnames(data)[12] <- c("ngaps")
colnames(data)[13] <- c("pctgaps")
rm(gaps)

# (3) REORDER, WRITE OUTPUT ----------------------------------------------------
data <- data %>% select(protein, gn, dom, all_para, all_non_ackr_para, all_cc_cxc_para, all_classa, 
                        cc_para, cxc_para, ack_para, ortho_cons, ngaps, pctgaps)

# write_csv(data, "data/sequence/gpcr/processed/CKR_CONSERVATION.csv") # LAST WRITTEN 20210328
