# Imports conservation scores generated by Mstatx and compiles them into a 
# single document. All Mstatx conservation calculations use alignments generated 
# from master alignment SEQUENCES_CK_ALL_ORTHO_PARA_NTERM_UPDATED.fasta located
# in folder /data/sequence/chemokine/alignments; CCN associated with alignment
# are in /data/lookup/ folder as SEQUENCES_CK_ALL_ORTHO_PARA_NTERM_UPDATED.txt. 
# Per position conservation scores by Mstatx were generated with the following 
# alignments sets (47 total):
# - Ortholog conservation (x43)
# - CC paralog conservation (1)
# - CXC paralog conservation (1)
# - CC and CXC paralog conservation (1)
# - All human (46 seq) paralog conservation (1)
# 
# Mstatx commands run using variations on the following command locally via
# command line terminal:
# mstatx -i ALL_para.fasta -s trident -m /Applications/MstatX-master/data/aaindex/HENS920102.mat -o ALL_para.txt
# (if doesn't work try going to /Users/akleist in terminal window (not R terminal) and running
# mstatx -i /Users/akleist/Desktop/Volkman_Lab/BABU_VOLKMAN/CHEMOKINE_PRJ/Data/chemokine_gpcr_encoding/data/sequence/chemokine/alignments/ALL_para.fasta -s trident -m /Applications/MstatX-master/data/aaindex/HENS920102.mat -o /Users/akleist/Desktop/Volkman_Lab/BABU_VOLKMAN/CHEMOKINE_PRJ/Data/chemokine_gpcr_encoding/data/sequence/chemokine/alignments/ALL_para.txt)
# (worked as of 12/22/23)
#
# Validated that scores are accurately assigned to columns via N-term spot
# checks - checked 0's appropriately assigned in data frame based on gapped
# columns from the alignment prior to the start of a given sequence

source("code/000_libraries.R")
source("code/000_functions.R")

################################################################################

# (1) IMPORT AND TIDY ----------------------------------------------------------
# import scores
all.para <- read.table("data/sequence/chemokine/mstatx/ALL_para.txt", header = FALSE)
all.para <- all.para %>% select(-V1)
colnames(all.para) <- c("all_para")

all.cc.cxc.para <- read.table("data/sequence/chemokine/mstatx/ALL_CC_CXC_para.txt", header = FALSE)
all.cc.cxc.para <- all.cc.cxc.para %>% select(-V1)
colnames(all.cc.cxc.para) <- c("all_cc_cxc_para")

cc.para <- read.table("data/sequence/chemokine/mstatx/CC_para.txt", header = FALSE)
cc.para <- cc.para %>% select(-V1)
colnames(cc.para) <- c("cc_para")

cxc.para <- read.table("data/sequence/chemokine/mstatx/CXC_para.txt", header = FALSE)
cxc.para <- cxc.para %>% select(-V1)
colnames(cxc.para) <- c("cxc_para")

ccl1.ortho <- read.table("data/sequence/chemokine/mstatx/ccl1_ortho.txt", header = FALSE)
ccl1.ortho <- ccl1.ortho %>% select(-V1)
colnames(ccl1.ortho) <- c("ccl1_ortho")

ccl2.ortho <- read.table("data/sequence/chemokine/mstatx/ccl2_ortho.txt", header = FALSE)
ccl2.ortho <- ccl2.ortho %>% select(-V1)
colnames(ccl2.ortho) <- c("ccl2_ortho")

ccl3.ortho <- read.table("data/sequence/chemokine/mstatx/ccl3_ortho.txt", header = FALSE)
ccl3.ortho <- ccl3.ortho %>% select(-V1)
colnames(ccl3.ortho) <- c("ccl3_ortho")

ccl3l1.ortho <- read.table("data/sequence/chemokine/mstatx/ccl3l1_ortho.txt", header = FALSE)
ccl3l1.ortho <- ccl3l1.ortho %>% select(-V1)
colnames(ccl3l1.ortho) <- c("ccl3l1_ortho")

ccl4.ortho <- read.table("data/sequence/chemokine/mstatx/ccl4_ortho.txt", header = FALSE)
ccl4.ortho <- ccl4.ortho %>% select(-V1)
colnames(ccl4.ortho) <- c("ccl4_ortho")

ccl5.ortho <- read.table("data/sequence/chemokine/mstatx/ccl5_ortho.txt", header = FALSE)
ccl5.ortho <- ccl5.ortho %>% select(-V1)
colnames(ccl5.ortho) <- c("ccl5_ortho")

ccl7.ortho <- read.table("data/sequence/chemokine/mstatx/ccl7_ortho.txt", header = FALSE)
ccl7.ortho <- ccl7.ortho %>% select(-V1)
colnames(ccl7.ortho) <- c("ccl7_ortho")

ccl8.ortho <- read.table("data/sequence/chemokine/mstatx/ccl8_ortho.txt", header = FALSE)
ccl8.ortho <- ccl8.ortho %>% select(-V1)
colnames(ccl8.ortho) <- c("ccl8_ortho")

ccl11.ortho <- read.table("data/sequence/chemokine/mstatx/ccl11_ortho.txt", header = FALSE)
ccl11.ortho <- ccl11.ortho %>% select(-V1)
colnames(ccl11.ortho) <- c("ccl11_ortho")

ccl13.ortho <- read.table("data/sequence/chemokine/mstatx/ccl13_ortho.txt", header = FALSE)
ccl13.ortho <- ccl13.ortho %>% select(-V1)
colnames(ccl13.ortho) <- c("ccl13_ortho")

ccl14.ortho <- read.table("data/sequence/chemokine/mstatx/ccl14_ortho.txt", header = FALSE)
ccl14.ortho <- ccl14.ortho %>% select(-V1)
colnames(ccl14.ortho) <- c("ccl14_ortho")

ccl15.ortho <- read.table("data/sequence/chemokine/mstatx/ccl15_ortho.txt", header = FALSE)
ccl15.ortho <- ccl15.ortho %>% select(-V1)
colnames(ccl15.ortho) <- c("ccl15_ortho")

ccl16.ortho <- read.table("data/sequence/chemokine/mstatx/ccl16_ortho.txt", header = FALSE)
ccl16.ortho <- ccl16.ortho %>% select(-V1)
colnames(ccl16.ortho) <- c("ccl16_ortho")

ccl17.ortho <- read.table("data/sequence/chemokine/mstatx/ccl17_ortho.txt", header = FALSE)
ccl17.ortho <- ccl17.ortho %>% select(-V1)
colnames(ccl17.ortho) <- c("ccl17_ortho")

ccl18.ortho <- read.table("data/sequence/chemokine/mstatx/ccl18_ortho.txt", header = FALSE)
ccl18.ortho <- ccl18.ortho %>% select(-V1)
colnames(ccl18.ortho) <- c("ccl18_ortho")

ccl19.ortho <- read.table("data/sequence/chemokine/mstatx/ccl19_ortho.txt", header = FALSE)
ccl19.ortho <- ccl19.ortho %>% select(-V1)
colnames(ccl19.ortho) <- c("ccl19_ortho")

ccl20.ortho <- read.table("data/sequence/chemokine/mstatx/ccl20_ortho.txt", header = FALSE)
ccl20.ortho <- ccl20.ortho %>% select(-V1)
colnames(ccl20.ortho) <- c("ccl20_ortho")

ccl21.ortho <- read.table("data/sequence/chemokine/mstatx/ccl21_ortho.txt", header = FALSE)
ccl21.ortho <- ccl21.ortho %>% select(-V1)
colnames(ccl21.ortho) <- c("ccl21_ortho")

ccl22.ortho <- read.table("data/sequence/chemokine/mstatx/ccl22_ortho.txt", header = FALSE)
ccl22.ortho <- ccl22.ortho %>% select(-V1)
colnames(ccl22.ortho) <- c("ccl22_ortho")

ccl23.ortho <- read.table("data/sequence/chemokine/mstatx/ccl23_ortho.txt", header = FALSE)
ccl23.ortho <- ccl23.ortho %>% select(-V1)
colnames(ccl23.ortho) <- c("ccl23_ortho")

ccl24.ortho <- read.table("data/sequence/chemokine/mstatx/ccl24_ortho.txt", header = FALSE)
ccl24.ortho <- ccl24.ortho %>% select(-V1)
colnames(ccl24.ortho) <- c("ccl24_ortho")

ccl25.ortho <- read.table("data/sequence/chemokine/mstatx/ccl25_ortho.txt", header = FALSE)
ccl25.ortho <- ccl25.ortho %>% select(-V1)
colnames(ccl25.ortho) <- c("ccl25_ortho")

ccl26.ortho <- read.table("data/sequence/chemokine/mstatx/ccl26_ortho.txt", header = FALSE)
ccl26.ortho <- ccl26.ortho %>% select(-V1)
colnames(ccl26.ortho) <- c("ccl26_ortho")

ccl27.ortho <- read.table("data/sequence/chemokine/mstatx/ccl27_ortho.txt", header = FALSE)
ccl27.ortho <- ccl27.ortho %>% select(-V1)
colnames(ccl27.ortho) <- c("ccl27_ortho")

ccl28.ortho <- read.table("data/sequence/chemokine/mstatx/ccl28_ortho.txt", header = FALSE)
ccl28.ortho <- ccl28.ortho %>% select(-V1)
colnames(ccl28.ortho) <- c("ccl28_ortho")

cxcl1.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl1_ortho.txt", header = FALSE)
cxcl1.ortho <- cxcl1.ortho %>% select(-V1)
colnames(cxcl1.ortho) <- c("cxcl1_ortho")

cxcl2.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl2_ortho.txt", header = FALSE)
cxcl2.ortho <- cxcl2.ortho %>% select(-V1)
colnames(cxcl2.ortho) <- c("cxcl2_ortho")

cxcl3.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl3_ortho.txt", header = FALSE)
cxcl3.ortho <- cxcl3.ortho %>% select(-V1)
colnames(cxcl3.ortho) <- c("cxcl3_ortho")

cxcl4.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl4_ortho.txt", header = FALSE)
cxcl4.ortho <- cxcl4.ortho %>% select(-V1)
colnames(cxcl4.ortho) <- c("cxcl4_ortho")

cxcl4l1.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl4l1_ortho.txt", header = FALSE)
cxcl4l1.ortho <- cxcl4l1.ortho %>% select(-V1)
colnames(cxcl4l1.ortho) <- c("cxcl4l1_ortho")

cxcl5.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl5_ortho.txt", header = FALSE)
cxcl5.ortho <- cxcl5.ortho %>% select(-V1)
colnames(cxcl5.ortho) <- c("cxcl5_ortho")

cxcl6.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl6_ortho.txt", header = FALSE)
cxcl6.ortho <- cxcl6.ortho %>% select(-V1)
colnames(cxcl6.ortho) <- c("cxcl6_ortho")

cxcl7.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl7_ortho.txt", header = FALSE)
cxcl7.ortho <- cxcl7.ortho %>% select(-V1)
colnames(cxcl7.ortho) <- c("cxcl7_ortho")

cxcl8.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl8_ortho.txt", header = FALSE)
cxcl8.ortho <- cxcl8.ortho %>% select(-V1)
colnames(cxcl8.ortho) <- c("cxcl8_ortho")

cxcl9.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl9_ortho.txt", header = FALSE)
cxcl9.ortho <- cxcl9.ortho %>% select(-V1)
colnames(cxcl9.ortho) <- c("cxcl9_ortho")

cxcl10.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl10_ortho.txt", header = FALSE)
cxcl10.ortho <- cxcl10.ortho %>% select(-V1)
colnames(cxcl10.ortho) <- c("cxcl10_ortho")

cxcl11.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl11_ortho.txt", header = FALSE)
cxcl11.ortho <- cxcl11.ortho %>% select(-V1)
colnames(cxcl11.ortho) <- c("cxcl11_ortho")

cxcl12.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl12_ortho.txt", header = FALSE)
cxcl12.ortho <- cxcl12.ortho %>% select(-V1)
colnames(cxcl12.ortho) <- c("cxcl12_ortho")

cxcl13.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl13_ortho.txt", header = FALSE)
cxcl13.ortho <- cxcl13.ortho %>% select(-V1)
colnames(cxcl13.ortho) <- c("cxcl13_ortho")

cxcl14.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl14_ortho.txt", header = FALSE)
cxcl14.ortho <- cxcl14.ortho %>% select(-V1)
colnames(cxcl14.ortho) <- c("cxcl14_ortho")

cxcl16.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl16_ortho.txt", header = FALSE)
cxcl16.ortho <- cxcl16.ortho %>% select(-V1)
colnames(cxcl16.ortho) <- c("cxcl16_ortho")

cxcl17.ortho <- read.table("data/sequence/chemokine/mstatx/cxcl17_ortho.txt", header = FALSE)
cxcl17.ortho <- cxcl17.ortho %>% select(-V1)
colnames(cxcl17.ortho) <- c("cxcl17_ortho")

cx3cl1.ortho <- read.table("data/sequence/chemokine/mstatx/cx3cl1_ortho.txt", header = FALSE)
cx3cl1.ortho <- cx3cl1.ortho %>% select(-V1)
colnames(cx3cl1.ortho) <- c("cx3cl1_ortho")

# combine
data <- bind_cols(all.para, all.cc.cxc.para, cc.para, cxc.para,
                  ccl1.ortho, ccl2.ortho, ccl3.ortho, ccl3l1.ortho, ccl4.ortho, 
                  ccl5.ortho, ccl7.ortho, ccl8.ortho, ccl11.ortho, ccl13.ortho,
                  ccl14.ortho, ccl15.ortho, ccl16.ortho, ccl17.ortho,
                  ccl18.ortho, ccl19.ortho, ccl20.ortho, ccl21.ortho, ccl22.ortho,
                  ccl23.ortho, ccl24.ortho, ccl25.ortho, ccl26.ortho, ccl27.ortho,
                  ccl28.ortho, cxcl1.ortho, cxcl2.ortho, cxcl3.ortho, cxcl4.ortho,
                  cxcl4l1.ortho, cxcl5.ortho, cxcl6.ortho, cxcl7.ortho, cxcl8.ortho,
                  cxcl9.ortho, cxcl10.ortho, cxcl11.ortho, cxcl12.ortho,
                  cxcl13.ortho, cxcl14.ortho, cxcl16.ortho, cxcl17.ortho, cx3cl1.ortho)

rm(all.para, all.cc.cxc.para, cc.para, cxc.para,
   ccl1.ortho, ccl2.ortho, ccl3.ortho, ccl3l1.ortho, ccl4.ortho, 
   ccl5.ortho, ccl7.ortho, ccl8.ortho, ccl11.ortho, ccl13.ortho,
   ccl14.ortho, ccl15.ortho, ccl16.ortho, ccl17.ortho,
   ccl18.ortho, ccl19.ortho, ccl20.ortho, ccl21.ortho, ccl22.ortho,
   ccl23.ortho, ccl24.ortho, ccl25.ortho, ccl26.ortho, ccl27.ortho,
   ccl28.ortho, cxcl1.ortho, cxcl2.ortho, cxcl3.ortho, cxcl4.ortho,
   cxcl4l1.ortho, cxcl5.ortho, cxcl6.ortho, cxcl7.ortho, cxcl8.ortho,
   cxcl9.ortho, cxcl10.ortho, cxcl11.ortho, cxcl12.ortho,
   cxcl13.ortho, cxcl14.ortho, cxcl16.ortho, cxcl17.ortho, cx3cl1.ortho)

# assign column names
aln.names <- c(read.table("data/lookup/SEQUENCES_CK_ALL_ORTHO_PARA_NTERM_UPDATED.txt", sep = ",", colClasses = "character"))
aln.names <- as.data.frame(aln.names)
aln.names <- t(aln.names)
aln.names <- as.data.frame(aln.names)
colnames(aln.names) <- c("ccn")
data <- bind_cols(aln.names, data)
rownames(data) <- 1:nrow(data)
rm(aln.names)

# spread data
data <- data %>% gather(protein, ortho_cons, 6:ncol(data))


# (2) ADD ADDITIONAL DESCRIPTORS -----------------------------------------------
# add domain
lookup <- read_csv("data/lookup/lookup_gnccn_to_domain.csv")
lookup <- lookup %>% unique()
data$dom <- lookup$dom[match(unlist(data$ccn), lookup$bwccn)]
rm(lookup)

# add gap numbers (CKR orthologs; 951 seqs) and percentages
aln <- readAAMultipleAlignment("data/sequence/chemokine/alignments/SEQUENCES_CK_ALL_ORTHO_PARA_NTERM_UPDATED.fasta")
aln <- as.matrix(aln)
aln.names <- c(read.table("data/lookup/SEQUENCES_CK_ALL_ORTHO_PARA_NTERM_UPDATED.txt", sep = ",", colClasses = "character"))
colnames(aln) <- as.matrix(aln.names)
aln <- as.data.frame(aln)
#rm(aln.names)

gaps <- NULL
for (i in 1:ncol(aln)){
  temp <- as.data.frame(table(aln[,i]))
  colnames(temp) <- c("resno", "n")
  temp <- subset(temp, resno == "-")
  temp$pos <- i
  gaps <- rbind.data.frame(gaps, temp)
  rm(temp)
}
rm(i)
gaps$pct <- gaps$n/nrow(aln)
aln.names <- t(as.data.frame(aln.names))
colnames(aln.names) <- c("ccn")
rm(aln)

gaps <- bind_cols(aln.names, gaps)
rm(aln.names)
colnames(gaps) <- c("ccn", "restype", "n", "pos", "pct")
gaps <- select(gaps, ccn, n, pct)
gaps$ccn <- as.character(gaps$ccn)

data <- left_join(data, gaps)
colnames(data)[9] <- c("ngaps")
colnames(data)[10] <- c("pctgaps")
rm(gaps)

# clean chemokine name
data <- data %>% separate(protein, into = c("protein", "b"), sep = "_") %>%
  dplyr::select(-b)
data$protein <- toupper(data$protein)

# (3) REORDER, WRITE OUTPUT ----------------------------------------------------
data <- data %>% dplyr::select(protein, ccn, dom, all_para, all_cc_cxc_para,
                        cc_para, cxc_para, ortho_cons, ngaps, pctgaps)

# write_csv(data, "data/sequence/chemokine/processed/CK_CONSERVATION.csv") # LAST WRITTEN 20231222

